src/hypothesis/strategies.py

```diff
@@ -1005,39 +1005,91 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
+    if max_denominator is not None:
+        if max_denominator < 1:
+            raise InvalidArgument(
+                'max_denominator=%r must be >= 1' % max_denominator)
+
+        def fraction_bounds(value):
+            """Find the best lower and upper approximation for value."""
+            # Adapted from CPython's Fraction.limit_denominator here:
+            # https://github.com/python/cpython/blob/3.6/Lib/fractions.py#L219
+            if value is None or value.denominator <= max_denominator:
+                return value, value
+            p0, q0, p1, q1 = 0, 1, 1, 0
+            n, d = value.numerator, value.denominator
+            while True:
+                a = n // d
+                q2 = q0 + a * q1
+                if q2 > max_denominator:
+                    break
+                p0, q0, p1, q1 = p1, q1, p0 + a * p1, q2
+                n, d = d, n - a * d
+            k = (max_denominator - q0) // q1
+            low, high = Fraction(p1, q1), Fraction(p0 + k * p1, q0 + k * q1)
+            assert low < value < high
+            return low, high
+
+        # Take the high approximation for min_value and low for max_value
+        bounds = (max_denominator, min_value, max_value)
+        _, min_value = fraction_bounds(min_value)
+        max_value, _ = fraction_bounds(max_value)
+
+        if None not in (min_value, max_value) and min_value > max_value:
+            raise InvalidArgument(
+                'There are no rational numbers with a denominator <= %r '
+                'between min_value=%r and max_value=%r' % bounds)
+
+    if min_value is not None and min_value == max_value:
+        return just(min_value)
 
     def dm_func(denom):
-        max_num = max_value * denom if max_value is not None else None
-        min_num = min_value * denom if min_value is not None else None
+        """Take denom, construct numerator strategy, and build fraction."""
+        # Four cases of algebra to get integer bounds and scale factor.
+        min_num, max_num = None, None
+        if max_value is None and min_value is None:
+            pass
+        elif min_value is None:
+            max_num = denom * max_value.numerator
+            denom *= max_value.denominator
+        elif max_value is None:
+            min_num = denom * min_value.numerator
+            denom *= min_value.denominator
+        else:
+            min_num = min_value.numerator * denom * max_value.denominator
+            max_num = max_value.numerator * denom * min_value.denominator
+            denom *= min_value.denominator * max_value.denominator
 
         return builds(
             Fraction,
             integers(min_value=min_num, max_value=max_num),
             just(denom)
         )
 
-    return denominator_strategy.flatmap(dm_func)
+    if max_denominator is None:
+        return integers(min_value=1).flatmap(dm_func)
+
+    return integers(1, max_denominator).flatmap(dm_func)\
+        .map(lambda f: f.limit_denominator(max_denominator))
```

Comment:
Nit: Stylistically I prefer to have `.map(` on the first line rather than an explicit line continuation.